### PR TITLE
feat(move): #DRIV-19 move folders from WS to NC is now working

### PR DESCRIPTION
**Pull request by aymericdevriesere**
## Describe your changes
Following #28 We are now able to drop folders from workspace to nextcloud.
We added the corresponding changes to the API and added the copy API too.
Documentation: https://entconf.gdapublic.fr/pages/viewpage.action?spaceKey=DRIV&title=Nextcloud+Documents+API
## Checklist tests

- [x] Drop one folder
- [x] Drop one file
- [x] Drop multiples folders
- [x] Drop multiples files
- [x] Drop multiples files and folders
- [x] Drop multiples files and folders with spaces and special characters in their names

## Issue ticket number and link
[ DRIV - 19 ]
https://entsupport.gdapublic.fr/projects/DRIV/issues/DRIV-19
## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/java/fr/openent/nextcloud/service/impl/DefaultDocumentsService.java [hunk 1]
```diff
@@ -725,53 +746,166 @@ public Future<JsonObject> uploadFile(UserNextcloud.TokenProvider user, Attachmen
...
-        Future<Void> current = Future.succeededFuture();
-        JsonObject result = new JsonObject();
+        JsonArray results = new JsonArray();
+        Future<JsonObject> current = Future.succeededFuture();
         for (String id : idList) {
-            current = current.compose(v -> moveFromWorkspaceToNC(userSession, id, parentName, result));
+            current = current.compose(v -> {
+                Future<JsonObject> future = handleDocumentCopy(userSession, user, id, parentName);
+                Promise<JsonObject> promiseSucceed = Promise.promise();
+                future.onComplete(moveStatus -> {
```
Comment (BastienBenard) [inline src/main/java/fr/openent/nextcloud/service/impl/DefaultDocumentsService.java:765]: plutot faire un 
onSuccess -> results.add(moveStatus.result());
onFail -> results.add(new JsonObject().put("error", err.getMessage));

File: src/main/java/fr/openent/nextcloud/service/impl/DefaultDocumentsService.java [hunk 2]
```diff
@@ -725,53 +746,166 @@ public Future<JsonObject> uploadFile(UserNextcloud.TokenProvider user, Attachmen
...
+     * @return Future with the details of the copy
+     */
+    private Future<JsonObject> handleFolderCopy(UserNextcloud.TokenProvider userSession, UserInfos user, JsonObject document, String parentPath) {
+        Promise<JsonObject> promise = Promise.promise();
+        JsonObject folderData = new JsonObject();
+        JsonObject list = new JsonObject()
+                .put(Field.ACTION, WorkspaceEventBusActions.LIST.action())
+                .put(Field.USERID_CAPS, userSession.userId())
+                .put(Field.PARENTID, document.getString(Field.UNDERSCORE_ID));
+        getUniqueFileName(userSession, (parentPath != null ? parentPath + "/" : "") + document.getString(Field.NAME), 0)
```
Comment (BastienBenard) [inline src/main/java/fr/openent/nextcloud/service/impl/DefaultDocumentsService.java:889]: On peut utilisé un constante URL_SEPARATOR = "/"

**Review by JinDouang — CHANGES_REQUESTED** [review]

File: src/main/java/fr/openent/nextcloud/service/impl/DefaultDocumentsService.java [hunk 3]
```diff
@@ -725,53 +746,166 @@ public Future<JsonObject> uploadFile(UserNextcloud.TokenProvider user, Attachmen
...
+     */
+    private Future<JsonObject> handleDocumentMove(UserNextcloud.TokenProvider userSession, UserInfos user, String id, String parentPath) {
+        Promise<JsonObject> promise = Promise.promise();
+        handleDocumentCopy(userSession, user, id, parentPath)
+                .onSuccess(res -> {
+                    JsonObject delete =  new JsonObject()
+                            .put(Field.ACTION, WorkspaceEventBusActions.DELETE.action())
+                            .put(Field.ID, id)
+                            .put(Field.USERID_CAPS, userSession.userId());
+                    EventBusHelper.requestJsonArray(eventBus, delete);
```
[the remark this answers is not in the document]
Comment (aymericdevriesere) [reply src/main/java/fr/openent/nextcloud/service/impl/DefaultDocumentsService.java:863]: On ne veut pas attendre le retour, on lance l'appel pour que le fichier soit supprimer mais le résultat importe peu car on ne fera aucune action en cas d'échec.

File: src/main/java/fr/openent/nextcloud/helper/EventBusHelper.java [hunk 4]
```diff
@@ -14,25 +14,6 @@
...
     private static final String WORKSPACE_BUS_ADDRESS = "org.entcore.workspace";
-    private static final Logger log = LoggerFactory.getLogger(DefaultDocumentsService.class);
-
-    /**
-     * Create a folder ine the workspace
-     * @param eb                The current eventBus
-     * @param action            The action parameters to send to the event bus
-     * @return                  Future with the status of the folder creation
-     */
-    public static Future<JsonObject> createFolder(EventBus eb, JsonObject action) {
```
[the remark this answers is not in the document]
Comment (aymericdevriesere) [reply src/main/java/fr/openent/nextcloud/helper/EventBusHelper.java:25]: J'ai supprimé toutes les fonctions de l'eventBus car elles ne servaient plus

File: src/main/java/fr/openent/nextcloud/helper/EventBusHelper.java [hunk 5]
```diff
@@ -14,25 +14,6 @@
...
-        return requestJsonObject(eb, action);
-    }
-
-    public static Future<JsonArray> deleteDocument(EventBus eb, String id, String userId) {
```
[the remark this answers is not in the document]
Comment (aymericdevriesere) [reply src/main/java/fr/openent/nextcloud/helper/EventBusHelper.java:29]: J'ai supprimé toutes les fonctions de l'eventBus car elles ne servaient plus

**Review by JinDouang — APPROVED** [review]

**Pull request merged**